### PR TITLE
RD-1291 Introduced TS to backend

### PR DESCRIPTION
**Pull request by kubama**
## Description

Although this PR is massive it's not as bad as it looks. All changes made to production and test code is just migration to TS modules, so the set of applied changes is pretty consistent across these files. Additionally there are some required changes around modules mocking in jest tests - I can elaborate about these on our weekly, as there are some gotchas.
Core changes are done in just a few files - `package.json`, `tsconfig.json` and `.eslintrc.json`.
Unfortunately, a few files that were properly renamed appear as deleted/added after committing 🙁. 

## Screenshots / Videos
N/A

## Checklist

- [X] My code follows the UI Code Style.
- [x] I verified that all tests and checks have passed.
- [X] I verified if that change requires a documentation update.
- [X] I added proper labels to this PR.

## Tests
https://jenkins.cloudify.co/view/UI/job/Stage-UI-System-Test/1185/

## Documentation
N/A

### CODE REVIEW

File: backend/routes/Filters.ts [hunk 1]
```diff
@@ -1,23 +1,15 @@
...
 const router = express.Router();
 
 router.use(passport.authenticate('token', { session: false }));
 
-router.use('/:service', (req, res, next) => {
-    logger.debug(
-        `${req.method} request on service '${req.params.service}' called with parameters: ${JSON.stringify(req.query)}`
-    );
-    return callService(req.params.service, req.method, req, res, next).catch(next);
+router.get('/usage/:filterId', (req, res, next) => {
```
Comment (stepek) [inline backend/routes/Filters.ts:9]: could you explain this change?

**Comment by stepek** [comment]
@kubama can you split #d447df2 to smaller commits?

File: backend/routes/Filters.ts (hunk 1 above)
Comment (kubama) [reply backend/routes/Filters.ts:9]: I did a mistake here and it was corrected

**Comment by kubama** [comment]
> @kubama can you split #d447df2 to smaller commits?

I could, but I'm not sure it would be of much help. As I stated in the description, the core changes are limited to just a few json files. All the changes to the actual source files are just conversion to TS modules. There are no changes to a single file that could be split across different commits.

**Comment by qooban** [comment]
Oh, quite big change. It'll take some time to review.

> Unfortunately, a few files that were properly renamed appear as deleted/added after committing slightly_frowning_face.

Having one commit containing both - rename and migration-related changes - won't help. AFAIR from what Grzegorz found is that GitHub works better if you do the rename in one commit and the changes inside the file in the other commit.

> https://jenkins.cloudify.co/view/UI/job/Stage-UI-System-Test/1177/

I see there's a problem with starting the backend:
```
14:15:02  > stage-backend@6.3.0--.dev1 wait-on-server /opt/cloudify-stage/backend
14:15:02  > wait-on tcp:localhost:8088 --timeout 5000 && echo 'Stage backend is up!'
14:15:02  
14:15:07  Error: Timeout
14:15:07      at Timeout._onTimeout (/opt/cloudify-stage/backend/node_modules/wait-on/lib/wait-on.js:113:10)
14:15:07      at listOnTimeout (internal/timers.js:554:17)
14:15:07      at processTimers (internal/timers.js:497:7)
14:15:07  npm ERR! code ELIFECYCLE
14:15:07  npm ERR! errno 1
14:15:07  npm ERR! stage-backend@6.3.0--.dev1 wait-on-server: `wait-on tcp:localhost:8088 --timeout 5000 && echo 'Stage backend is up!'`
14:15:07  npm ERR! Exit status 1
14:15:07  npm ERR! 
14:15:07  npm ERR! Failed at the stage-backend@6.3.0--.dev1 wait-on-server script.
14:15:07  npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```
Do we need any update in terms of installed packages on the server to execute `ts-node`? Maybe it should be installed globally? (`npm i -g ts-node`)

Other questions before I jump into the review (possibly tomorrow):
1. Did you manage to change all the files to `.ts`?
2. How many of them are already fully converted?
3. Could you provide some guidelines on how to do similar migration in Composer? (it'd be very much appreciated)

**Comment by kubama** [comment]
> Having one commit containing both - rename and migration-related changes - won't help. AFAIR from what Grzegorz found is that GitHub works better if you do the rename in one commit and the changes inside the file in the other commit.

Yes, I didn't expect that as I did just minor updates to the renamed files. In fact git respected renames for most of the files.

> I see there's a problem with starting the backend:
> 
> ```
> 14:15:02  > stage-backend@6.3.0--.dev1 wait-on-server /opt/cloudify-stage/backend
> 14:15:02  > wait-on tcp:localhost:8088 --timeout 5000 && echo 'Stage backend is up!'
> 14:15:02  
> 14:15:07  Error: Timeout
> 14:15:07      at Timeout._onTimeout (/opt/cloudify-stage/backend/node_modules/wait-on/lib/wait-on.js:113:10)
> 14:15:07      at listOnTimeout (internal/timers.js:554:17)
> 14:15:07      at processTimers (internal/timers.js:497:7)
> 14:15:07  npm ERR! code ELIFECYCLE
> 14:15:07  npm ERR! errno 1
> 14:15:07  npm ERR! stage-backend@6.3.0--.dev1 wait-on-server: `wait-on tcp:localhost:8088 --timeout 5000 && echo 'Stage backend is up!'`
> 14:15:07  npm ERR! Exit status 1
> 14:15:07  npm ERR! 
> 14:15:07  npm ERR! Failed at the stage-backend@6.3.0--.dev1 wait-on-server script.
> 14:15:07  npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
> ```
> 
> Do we need any update in terms of installed packages on the server to execute `ts-node`? Maybe it should be installed globally? (`npm i -g ts-node`)

Nothing extra needs to be done.

The issue here seems to be just the timeout - the backend now starts slightly longer, which is not unexpected. I increased the timeout and submitted new job (link in description updated).

> 1. Did you manage to change all the files to `.ts`?

No, db migrations and sequelize models (also used by migrations) are not changed. These are not directly imported, and the logic that handles them is part of cloudify-ui-common and used by composer as well - if we want these to be converted as well it will be best to first migrate composer backend to TS and then handle it as part of a separate ticket.

> 2. How many of them are already fully converted?

6 production source files and 3 test files are fully migrated.

> 3. Could you provide some guidelines on how to do similar migration in Composer? (it'd be very much appreciated)

Here are the steps in short:
1. Install `ts-node` and `typescript` as production dependencies
2. Rename all source files (except migrations and sequelize models) and test files to .ts
2. Replace all `node` occurrences with `ts-node` and all direct references to js files with their corresponding ts files
3. Look for all wildcards referencing js files and change them so they referenced both js and ts files
3. Convert all renamed files to valid TS modules (import/export)
4. Test files that used `require` inside functions will need to have module mocking updated:
- replace all `jest.doMock` calls with `jest.mock`
- move all `jest.mock` calls to top level
- make sure all required `jest.mock` calls are done directly in the test file (not in an imported file)
7. Align eslint configuration

**Comment by kubama** [comment]
There is one system test failure - looks like an issue with widget backend. I'll investigate.

File: backend/package.json [hunk 2]
```diff
@@ -59,11 +59,17 @@
...
+    "typescript": "^4.3.5",
     "vm2": "^3.5.2",
     "wait-on": "^3.3.0"
   },
   "devDependencies": {
+    "@babel/preset-typescript": "^7.15.0",
+    "@types/fs-extra": "^9.0.12",
+    "@types/lodash": "^4.14.172",
     "@types/node": "^12.20.1",
+    "@types/passport": "^1.0.7",
```
Comment (qooban) [inline backend/package.json:72]: I'm wondering - is RPM package is usable with those changes? 
(it does not include `devDependencies`, see `%prep` step in `packaging/cloudify-stage.spec`)

File: backend/serverSettings.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                const mode = process.argv[index + 1].toLowerCase();
+                if (_.includes(modes, mode)) {
+                    serverMode = mode;
+                } else {
+                    displayUsage();
+                }
+            } else {
+                displayUsage();
+            }
+        }
```
Comment (qooban) [inline backend/serverSettings.ts:44]: Is it working correctly?
I see `-mode` switch in the original file.

We need to have main and community modes working, I haven't seen anyone using customer mode.

File: backend/test/.eslintrc.json [hunk 4]
```diff
@@ -5,17 +5,17 @@
     "extends": [
         "plugin:jest/recommended"
     ],
-    "settings": {
-        "import/resolver": {
-            "jest": {
-                "jestConfigFile": "jest.config.js"
-            }
-        }
-    },
```
Comment (qooban) [inline backend/test/.eslintrc.json:14]: What was the reason behind removing it?

Did you had problems with resolving `.ts` files? (if so, see my comment: https://github.com/cloudify-cosmo/cloudify-blueprint-composer/pull/921/files#r700138908)

File: backend/test/routes/Config.test.ts [hunk 5]
```diff
@@ -1,8 +1,11 @@
...
-const request = require('supertest');
-const app = require('app');
-const appConfig = require('conf/config.json');
-const userConfig = require('conf/userConfig.json');
+// @ts-nocheck File not migrated fully to TS
+import _ from 'lodash';
+import request from 'supertest';
+import app from 'app';
+import appConfig from '../../../conf/config.json';
+import userConfig from '../../../conf/userConfig.json';
```
Comment (qooban) [inline backend/test/routes/Config.test.ts:6]: If we bring back jest import resolver, we can avoid using such paths.

File: backend/tsconfig.json [hunk 6]
```diff
@@ -1,5 +1,7 @@
 {
-    "extends": "../tsconfig.base.json",
-    "exclude": ["node_modules", "test", "dist"],
-    "include": ["**/*", "../conf/**/*.json"]
+    "extends": "ts-node/node12/tsconfig.json",
```
Comment (qooban) [inline backend/tsconfig.json:2]: We should possibly add a point to https://cloudifysource.atlassian.net/wiki/spaces/RD/pages/682917889/Upgrading+Node.js about it.

File: backend/tsconfig.json [hunk 7]
```diff
@@ -1,5 +1,7 @@
 {
-    "extends": "../tsconfig.base.json",
-    "exclude": ["node_modules", "test", "dist"],
```
Comment (qooban) [inline backend/tsconfig.json:3]: No need for excluding `node_modules`?

File: devServer.js [hunk 8]
```diff
@@ -6,13 +6,12 @@ const webpack = require('webpack');
...
 
 const webpackConfig = require('./webpack.config')({}, { mode: 'development' });
-const Consts = require('./backend/consts');
 const startWidgetBackendWatcher = require('./scripts/widgetBackendWatcher');
 
 const host = 'localhost';
 const devServerPort = 4000;
 const stageBackendPort = 8088;
-const contextPath = Consts.CONTEXT_PATH;
+const contextPath = '/console';
```
Comment (qooban) [inline devServer.js:14]: I guess it's due to the fact that `./backend/consts` is TS file.
Will it be possible to have TS support for webpack dev server? (can be done within separate ticket of course, if not trivial)

**Review by qooban — COMMENTED** [review]
I did partial review (78/124 files) so far. Will continue later, probably tomorrow (due to the planned meetings). Added some comments.

Very hard to review files which are detected as deleted/added. It's not the time right now, but I'd suggest for the future to consider splitting such huge PR into smaller pieces. I suspect it must be possible to have it done within few PRs, one building on top of the other.

File: backend/package.json [hunk 9]
```diff
@@ -5,24 +5,24 @@
...
+    "db-migrate-down-to": "ts-node migration downTo",
+    "db-migrate-reset": "ts-node migration reset",
+    "db-migrate-status": "ts-node migration status",
+    "startWithNodemon": "cross-env NODE_ENV=development NODE_TLS_REJECT_UNAUTHORIZED='0' nodemon -x ts-node -e js,ts,json --watch . --watch ../conf",
+    "devDebug": "npm run startWithNodemon -- --inspect server.ts",
+    "devStart": "npm run startWithNodemon -- server.ts",
+    "devTrace": "npm run startWithNodemon -- --trace-warnings server.ts",
     "postinstall": "patch-package",
-    "start": "node server.js ${STAGE_BACKEND_ARGS}",
+    "start": "ts-node server.ts ${STAGE_BACKEND_ARGS}",
```
[the remark this answers is not in the document]
Comment (kubama) [reply backend/package.json:23]: Type checking is done as part of every CI build, so I don't see much risk here. But I understand the concern, I think adding the `--transpile-only` for extra safety is a good idea, I'll add it.

> alternative is running tsc first, shipping backend transpiled already and then do no changes in npm run start

I imagine this could be the next step addressed in a separate ticket as it would complicate things considerably more - this task is already massive and went far beyond its estimate.

File: backend/serverSettings.ts (hunk 3 above)
Comment (kubama) [reply backend/serverSettings.ts]: Good catch, no idea how it got changed. Fixed.

File: devServer.js (hunk 8 above)
Comment (kubama) [reply devServer.js:14]: I have no idea, I didn't investigate such possibility.

File: backend/tsconfig.json (hunk 7 above)
Comment (kubama) [reply backend/tsconfig.json:3]: I'm not sure but I don't see any difference.

File: backend/tsconfig.json (hunk 6 above)
Comment (kubama) [reply backend/tsconfig.json:2]: I think the target placement for this is in `cloudify-ui-common` once composer backend is migrated. So I'd wait with documentation updates for finalization.

File: backend/test/routes/Config.test.ts (hunk 5 above)
Comment (kubama) [reply backend/test/routes/Config.test.ts:6]: Not really - the import resolver impacts eslint only. Here the issue was tsc not resolving the modules as they were originally specified

File: backend/test/.eslintrc.json (hunk 4 above)
Comment (kubama) [reply backend/test/.eslintrc.json:14]: Yes, I couldn't make it working.
In the end there are two rules that check the same thing: `node/no-missing-import` and `import/no-unresolved`. There is no point in having both, so I decided to disable the second one (thus removing the setting).

File: backend/.eslintrc.json [hunk 10]
```diff
@@ -1,8 +1,17 @@
 {
```
[the remark this answers is not in the document]
Comment (kubama) [reply backend/.eslintrc.json:1]: I think this should be done as part of introducing TS to composer backend

File: backend/package.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (kubama) [reply backend/package.json]: > I'll add it.

Done. Checked and it works as expected.

File: backend/babel.config.js [hunk 11]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+    presets: [['@babel/preset-env', { targets: { node: 'current' } }], '@babel/preset-typescript']
```
[the remark this answers is not in the document]
Comment (kubama) [reply backend/babel.config.js:2]: From babel docs:
> If you want to compile against the current node version, you can specify "node": "current", which would be the same as "node": process.versions.node.

File: backend/package.json (hunk 2 above)
Comment (kubama) [reply backend/package.json:72]: Looks like this is where the just added `--transpile-only` (`-T`) flag is crucial. After adding it the typings are no longer required for `ts-node` to run, so it works without them.

**Comment by kubama** [comment]
> It's not the time right now, but I'd suggest for the future to consider splitting such huge PR into smaller pieces. I suspect it must be possible to have it done within few PRs, one building on top of the other.

Thinking of it now I think it could be split into first introducing TS support without renaming any of the source files, and then gradually renaming the files and applying required changes (TS modules). Unfortunately I wasn't upfront aware of the scale of the changes that were required. I hope it's a lesson learned for composer backend work.

File: devServer.js (hunk 8 above)
Comment (stepek) [reply devServer.js:14]: +1 to move this to const files.

Comment (kubama) [reply devServer.js:14]: OK, I'll give it a try

File: backend/babel.config.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (qooban) [reply backend/babel.config.js:2]: I see. And according to https://nodejs.org/api/process.html#process_process_versions and my quick local tests it seems that `process.versions.node` is a version of node used to start the process.

All good then 👍🏻

**Review by stepek — APPROVED** [review]
any new comments from me.

File: backend/package.json (hunk 2 above)
Comment (qooban) [reply backend/package.json:72]: Great 👍🏻

File: backend/samlSetup.ts [hunk 12]
```diff
@@ -1,8 +1,5 @@
-/**
- * Created by edenp on 07/09/2017.
- */
-
-exports.validate = samlConfig => {
+// @ts-nocheck File not migrated fully to TS
+export default samlConfig => {
```
Comment (qooban) [inline backend/samlSetup.ts:2]: That change in the export name will introduce a bug, as in `app.ts` we call `samlSetup.validate`.

File: backend/handler/services/LoggerService.ts [hunk 13]
```diff
@@ -0,0 +1,3 @@
+import { getLogger } from '../LoggerHandler';
+
+export default (category = '') => getLogger(`WidgetBackend${category ? `-${category}` : ''}`);
```
Comment (qooban) [inline backend/handler/services/LoggerService.ts:3]: I see a change in the export, so will it work properly with our widget backends (Executions and Spire Manager widgets)?

File: backend/handler/services/ManagerService.ts [hunk 14]
```diff
@@ -0,0 +1,56 @@
+// @ts-nocheck File not migrated fully to TS
+/* eslint no-underscore-dangle: ["error", { "allow": ["_size", "_offset"] }] */
+
+import _ from 'lodash';
+import param from 'jquery-param';
```
Comment (qooban) [inline backend/handler/services/ManagerService.ts:5]: Oh, we still use jquery in some way :-) 

Can we convert it at once or have a ticket for removing it?

@stepek FYI.

File: backend/test/migration.test.ts [hunk 15]
```diff
@@ -1,16 +1,21 @@
...
 const latestMigration = '20210519093609-6_0-UserAppsManagerIpColumnRemoval.js';
-const userTemplatesFolder = Utils.getResourcePath('templates', true);
+const userTemplatesFolder = getResourcePath('templates', true);
 const userTemplatesBackupFolder = `${userTemplatesFolder}-backup`;
 
 describe('Migration script', () => {
-    beforeAll(() => execSync('node migration.js up'));
+    function execMigration(command: string) {
+        return execSync(`ts-node migration ${command}`);
+    }
```
Comment (qooban) [inline backend/test/migration.test.ts:16]: Good idea to introduce this function 👍🏻 

I'm taking a note to do the same in Composer (RD-3044).

File: backend/test/handlers/services/LoggerService.test.ts [hunk 16]
```diff
@@ -0,0 +1,14 @@
+import { getLogger } from 'handler/LoggerHandler';
+import Logger from 'handler/services/LoggerService';
+
+jest.mock('handler/LoggerHandler');
+const loggerMock = {};
+(<jest.Mock>getLogger).mockReturnValue(loggerMock);
+
+describe('LoggerService', () => {
```
Comment (qooban) [inline backend/test/handlers/services/LoggerService.test.ts:8]: Good idea to introduce tests for that service 👍🏻 

Side note: We've got a problem with logger (`winston` seems problematic) which we should solve soon - RD-2917 is planned for the next sprint.

File: backend/test/handlers/ArchiveHelper.test.ts [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+import { request } from 'handler/RequestHandler';
+
+jest.mock('handler/ManagerHandler');
+
+jest.mock('handler/RequestHandler', () => ({
+    request: jest.fn((method, url, headers, data, onSuccess, onError) => onError())
+}));
+
+describe('ArchiveHelper', () => {
+    it('fetches extenral data with correct headers', () => {
```
Comment (qooban) [inline backend/test/handlers/ArchiveHelper.test.ts:12]: Let's fix typo at once:
```suggestion
    it('fetches external data with correct headers', () => {
```

File: backend/test/handlers/ArchiveHelper.test.ts [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+        expect.assertions(1);
+        const url = 'http://wp';
+        return saveDataFromUrl(url).catch(() =>
```
Comment (qooban) [inline backend/test/handlers/ArchiveHelper.test.ts:15]: Good idea to add proper URL to check here 👍🏻

**Review by qooban — COMMENTED** [review]
Second part of the review. 104/124 files reviewed, so a bit closer to finish.

File: backend/config.ts [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+    userConfig = _.defaultsDeep(userDataConfig, userConfig); // Create full user configuration
+} catch (err) {
+    if (err.code !== 'MODULE_NOT_FOUND') {
+        throw err;
+    }
+}
+
+let me = null;
+
+export function loadMeJson() {
```
Comment (qooban) [inline backend/config.ts:31]: Name of this function sounds funny, but I don't have better one, so we can keep it 🙂

**Review by qooban — COMMENTED** [review]
OK, I'm done with the review. I have no more additional comments. I didn't manage to review all the files with the same attention to detail, but I believe that we have some time until the next release to detect any potential problems.

File: devServer.js (hunk 8 above)
Comment (kubama) [reply devServer.js]: Looks like it was not that difficult - done.

File: backend/package.json [hunk 20]
```diff
@@ -5,24 +5,24 @@
...
-    "devStart": "npm run startWithNodemon -- server.js",
-    "devTrace": "npm run startWithNodemon -- --trace-warnings server.js",
+    "db-migrate-current": "ts-node -T migration current",
+    "db-migrate-down-to": "ts-node -T migration downTo",
+    "db-migrate-reset": "ts-node -T migration reset",
+    "db-migrate-status": "ts-node -T migration status",
+    "startWithNodemon": "cross-env NODE_ENV=development NODE_TLS_REJECT_UNAUTHORIZED='0' nodemon -x 'ts-node -T' -e js,ts,json --watch . --watch ../conf",
+    "devDebug": "npm run startWithNodemon -- --inspect server.ts",
+    "devStart": "npm run startWithNodemon -- server.ts",
+    "devTrace": "npm run startWithNodemon -- --trace-warnings server.ts",
```
[the remark this answers is not in the document]
Comment (kubama) [reply backend/package.json:21]: I managed to make it working

File: backend/package.json [hunk 21]
```diff
@@ -5,24 +5,24 @@
...
-    "startWithNodemon": "cross-env NODE_ENV=development NODE_TLS_REJECT_UNAUTHORIZED='0' nodemon --watch . --watch ../conf",
-    "devDebug": "npm run startWithNodemon -- --inspect server.js",
-    "devStart": "npm run startWithNodemon -- server.js",
-    "devTrace": "npm run startWithNodemon -- --trace-warnings server.js",
+    "db-migrate-current": "ts-node -T migration current",
+    "db-migrate-down-to": "ts-node -T migration downTo",
+    "db-migrate-reset": "ts-node -T migration reset",
+    "db-migrate-status": "ts-node -T migration status",
+    "startWithNodemon": "cross-env NODE_ENV=development NODE_TLS_REJECT_UNAUTHORIZED='0' nodemon -x 'ts-node -T' -e js,ts,json --watch . --watch ../conf",
+    "devDebug": "npm run startWithNodemon -- --inspect server.ts",
```
[the remark this answers is not in the document]
Comment (kubama) [reply backend/package.json:19]: I managed to make it working

File: backend/handler/services/LoggerService.ts (hunk 13 above)
Comment (kubama) [reply backend/handler/services/LoggerService.ts:3]: Yes, it's working correctly. It is covered by executions widget system test.

File: backend/samlSetup.ts (hunk 12 above)
Comment (kubama) [reply backend/samlSetup.ts:2]: Good catch - fixed. I also added small test to cover it.

File: backend/handler/services/ManagerService.ts (hunk 14 above)
Comment (kubama) [reply backend/handler/services/ManagerService.ts:5]: Actually this is a small package completely independent of jquery. Seems it's just the package name indicating that it does the same as `jQuery.param` does, nothing more.

File: package.json [hunk 22]
```diff
@@ -25,9 +25,9 @@
     "build:coverage": "cross-env COVERAGE_CHECK=1 npm run build",
     "ci:frontend": "npm ci --prefer-offline",
     "ci:backend": "npm ci --prefix backend --prefer-offline",
-    "check-types": "tsc --build",
+    "check-types": "tsc --build && cd backend && npm run check-types",
```
Comment (qooban) [inline package.json:28]: Optional:

```suggestion
    "check-types": "npm run check-types:frontend && npm run check-types:backend",
    "check-types:frontend": "tsc --build",
    "check-types:backend": "npm run --prefix backend check-types",
```

File: devServer.ts [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+import getWebpackConfig from './webpack.config';
+import startWidgetBackendWatcher from './scripts/widgetBackendWatcher';
+
+import { CONTEXT_PATH } from './backend/consts';
+
+const webpackConfig = getWebpackConfig({}, { mode: 'development' });
+
+const host = 'localhost';
+const devServerPort = 4000;
+const stageBackendPort = 8088;
```
Comment (qooban) [inline devServer.ts:13]: As you are referencing `backend/consts.ts` in this file, then we can possibly reuse more constants:
1. `SERVER_HOST`
2. `SERVER_PORT`

**Review by qooban — COMMENTED** [review]
Added two small comments.

File: backend/tsconfig.json (hunk 6 above)
Comment (qooban) [reply backend/tsconfig.json:2]: OK, I'll add a note in RD-3044.

File: backend/.eslintrc.json (hunk 10 above)
[the remark this answers is not in the document]
Comment (qooban) [reply backend/.eslintrc.json:1]: I'll add a not to RD-3044 and will see if I manage to tackle it.

File: devServer.ts (hunk 23 above)
Comment (kubama) [reply devServer.ts]: Done

**Review by qooban — APPROVED** [review]
I appreciate hard work on that migration 👍🏻 

Could you summarize leftovers (migrations? DB models? any other?) and possibly add tickets for them? (labelled with `UI stage typescript`)

**Comment by kubama** [comment]
@qooban Created RD-3119 and RD-3120
